go-runtime/ftl/map.go

```diff
@@ -21,3 +28,20 @@ func (mh *MapHandle[T, U]) Get(ctx context.Context) U {
 func Map[T, U any](getter Handle[T], fn func(context.Context, T) (U, error)) MapHandle[T, U] {
 	return MapHandle[T, U]{fn: fn, getter: getter}
 }
+
+// Mock takes a function that will get called in place of the logic that this Map was
+// instantiated with. It returns an unmock function to defer in tests:
+//
+//	unmock := map.Mock(...)
+//	defer unmock()
+//	_ := map.Mock(...)
+//	_ := map.Mock(...)
+//
+// Unmock restores the original behavior of the MapHandle no matter how many times it was
+// mocked in the meantime, so it only needs to be called once.
+func (mh *MapHandle[T, U]) Mock(mockFn func(context.Context) U) func() {
```

Comment:
We have a new system for differentiation between "real" code and "test" code should take place, the [FTL](https://github.com/TBD54566975/ftl/blob/main/go-runtime/internal/api.go#L14) interface. We should implement this in there somehow, and add options for `ftltest.Context()` to set it up and tear it down at the end of the test.

It should probably follow the pattern of how the fakes for verbs are done:

1. By default all `Map` calls under testing will panic
2. An option to allow all `Map` calls to pass through
3. A final option to mock individual calls.

These three will give full control.

Reply:
You'll need to do something like:

1. Add a `FTL.CallMap(ctx context.Context, mapper any) (any, error)` method to the `FTL` interface
1. Add `mockMaps` to `ftltest.OptionsState`
2. Extend `newFakeFTL()` to take in the `mockMaps`
3. Add options for `ftltest.Context()` to achieve the above
4. Replace the `ftl.Map.Get` implementation with one that uses the `FTL` interface

I think the API would be something like:

```go
// WhenMap injects a fake implementation of a Mapping function
func WhenMap[T, U any](mapper *ftl.MapHandle[T], fake func(context.Context, T) (U, error)) Option

// WithMapsAllowed allows all `ftl.Map` calls to pass through to their original implementation, rather than
func WithMapsAllowed() Option
```

Reply:
finally ready for review!